lean/bitvec.lean

```diff
@@ -0,0 +1,325 @@
+  -- bitwise negation
+  def not {w:ℕ} (x: bitvec w) : bitvec w := ⟨2^w - x.val - 1,
+    begin
+      have xval_pos : 0 < x.val + 1,
+      { apply (succ_pos x.val) },
+      apply (sub_lt _ xval_pos),
+      apply pos_pow_of_pos,
+      apply (succ_pos 1)
+    end⟩
+
+  -- logical bitwise and
+  def and {w:ℕ} (x y : bitvec w) : bitvec w := bitvec.of_nat w (nat.land x.val y.val)
+  -- logical bitwise or
+  def or  {w:ℕ} (x y : bitvec w) : bitvec w := bitvec.of_nat w (nat.lor  x.val y.val)
+  -- logical bitwise xor
+  def xor {w:ℕ} (x y : bitvec w) : bitvec w := bitvec.of_nat w (nat.lxor x.val y.val)
+
+  infix `.&&.`:70 := and
+  infix `.||.`:65 := or
+
+end bitwise
+
+section arith
+  -- Arithmetic operations on bitvectors
+  variable {n : ℕ}
+
+  -- Add with carry (no overflow)
+  def adc (x y : bitvec n) (c : bool) : bitvec n × bool :=
+    let c₁ := if c then 1 else 0,
+        r  := x.val + y.val + c₁ in
+    ⟨ bitvec.of_nat n r, r ≥ 2^n ⟩
+
+  protected def add (x y : bitvec n) : bitvec n := (adc x y ff).1
+
+  -- Subtract with borrow
+  def sbb (x y : bitvec n) (b : bool) : bool × bitvec n :=
+    let b₁ : bitvec n := if b then 1 else 0,
+        r  := match bitvec.adc x (bitvec.neg y) ff with
+              | (z, b₂) := bitvec.adc z (bitvec.neg b₁) ff
+              end
+    in ⟨ if b then y.val + 1 > x.val else y.val > x.val , r.1 ⟩
+
+  -- Usual arithmetic subtraction
+  protected def sub (x y : bitvec n) : bitvec n := (sbb x y ff).2
+
+  instance : has_add (bitvec n)  := ⟨bitvec.add⟩
+  instance : has_sub (bitvec n)  := ⟨bitvec.sub⟩
+  instance : has_neg (bitvec n)  := ⟨bitvec.neg⟩
+
+  protected def mul (x y : bitvec n) : bitvec n := bitvec.of_nat n (x.val * y.val)
+
+  instance : has_mul (bitvec n) := ⟨bitvec.mul⟩
+
+  def bitvec_pow (x: bitvec n) (k:ℕ) : bitvec n := bitvec.of_nat n (x.val^k)
+
+  instance bitvec_has_pow : has_pow (bitvec n) ℕ := ⟨bitvec_pow⟩
+
+end arith
+
+section shift
+  -- Shift related operations, including signed and unsigned shift.
+
+  variable {n : ℕ}
+
+  -- shift left
+  def shl (x : bitvec n) (i : ℕ) : bitvec n := bitvec.of_nat n (nat.shiftl x.val i)
+
+  -- unsigned shift right
+  def ushr (x : bitvec n) (i : ℕ) : bitvec n := bitvec.of_nat n (nat.shiftr x.val i)
+
+  -- signed shift right
+  def sshr (x: bitvec n) (i:ℕ) : bitvec n :=
+    match n with
+    | 0      := 0
+    | succ m := bitvec.of_int m (int.shiftr (bitvec.to_int x) i)
+    end
+
+end shift
+
+section listlike
+  -- Operations that treat bitvectors as a list of bits.
+
+  def lsb {n:ℕ} (x: bitvec n) : bool := x.val % 2 = 1
+  def nth {w:ℕ} (x : bitvec w) (idx : ℕ) : bool :=
+      x .&&. (shl 1 idx) ≠ 0
+
+  def zext {n:ℕ} (m:ℕ) (x: bitvec n) : bitvec (n+m) :=
+    ⟨ x.val, calc
+               x.val < 2^n     : x.property
+                 ... ≤ 2^(n+m) : begin
+                                   apply pow_le_pow_of_le_right, dec_trivial_tac,
+                                   apply le_add_right
+                                 end⟩
```

Comment:
I don't think `zext` is right.  It should be an signed extension, .e.g,

```
 def zext {n:ℕ} (m:ℕ) (x: bitvec n) : bitvec (n+m) := of_int (n+m) x.to_int
```

Reply:
Are you thinking of sign extension? I don't think zero extension (I can add a comment about the name), should treat the number as signed.

Reply:
I renamed `zext` to `uext` (@simonjwinwood take note, I think you were using it?) I changed the body to be analogous to the one you gave (but unsigned).

And added `sext` with the definition you gave.

Reply:
Oops, I think `uext` is more clear.  I interpreted the "z" in `zext` as int.